### PR TITLE
DevObj: fix race condition in work scheduling

### PR DESCRIPTION
**Pull request by bkueng**
see comments in the code.

### CODE REVIEW

File: framework/src/DevObj.cpp [hunk 1]
```diff
@@ -165,18 +166,22 @@ int DevObj::start(void)
...
-		} else {
-			return -4;
-		}
+			//if someone else calls getWorkHandle at the same time, both will return the same handle,
+			//and one of the schedule() calls will then fail with EBUSY. In that case just retry.
+		} while (m_work_handle.getError() == EBUSY);
 	}
 
-	return 0;
+	return -m_work_handle.getError();
```
Comment (julianoes) [inline framework/src/DevObj.cpp]: If there is no error, will `getError()` be 0?

File: framework/src/DevObj.cpp [hunk 2]
```diff
@@ -165,18 +166,22 @@ int DevObj::start(void)
...
-		if (m_work_handle.isValid()) {
 			DF_LOG_DEBUG("DevObj::start schedule %s", m_name);
 			WorkMgr::schedule(m_work_handle);
 
-		} else {
-			return -4;
-		}
+			//if someone else calls getWorkHandle at the same time, both will return the same handle,
+			//and one of the schedule() calls will then fail with EBUSY. In that case just retry.
+		} while (m_work_handle.getError() == EBUSY);
```
Comment (julianoes) [inline framework/src/DevObj.cpp]: Are you sure that we won't end up in this while loop forever?

File: framework/src/DevObj.cpp [hunk 3]
```diff
@@ -323,8 +328,19 @@ void DevObj::setSampleInterval(unsigned int sample_interval_usecs)
...
+
+				if (!m_work_handle.isValid()) {
+					return;
+				}
+
+				WorkMgr::schedule(m_work_handle);
+
+				//if someone else calls getWorkHandle at the same time, both will return the same handle,
+				//and one of the schedule() calls will then fail with EBUSY. In that case just retry.
+			} while (m_work_handle.getError() == EBUSY);
```
Comment (julianoes) [inline framework/src/DevObj.cpp]: Same here. Is `getError()` correctly reset?

File: framework/src/DevObj.cpp (hunk 2 above)
Comment (bkueng) [inline framework/src/DevObj.cpp]: Yes. If there are 2 threads one of them will succeed for sure, the other one will loop and succeed next time. Similarly for N threads.

File: framework/src/DevObj.cpp (hunk 1 above)
Comment (bkueng) [inline framework/src/DevObj.cpp]: Yes it's reset to 0 internally.

File: framework/src/DevObj.cpp (hunk 2 above)
Comment (julianoes) [inline framework/src/DevObj.cpp]: Ok I'll take your word :).

**Comment by julianoes** [comment]
@bkueng I assume you've tested this?

**Comment by bkueng** [comment]
yes. I never actually ran into the race problem, since it's fairly low prob, but I'm confident it's correct.

**Pull request merged**